Use restricted Damerau-Levenshtein distance for diagnostics

jhpratt opened the pull request:
This replaces the existing Levenshtein algorithm with the Damerau-Levenshtein algorithm. This means that "ab" to "ba" is one change (a transposition) instead of two (a deletion and insertion). More specifically, this is a _restricted_ implementation, in that "ca" to "abc" cannot be performed as "ca" → "ac" → "abc", as there is an insertion in the middle of a transposition. I believe that errors like that are sufficiently rare that it's not worth taking into account.

This was first brought up [on IRLO](https://internals.rust-lang.org/t/18227) when it was noticed that the diagnostic for `prinltn!` (transposed L and T) was `print!` and not `println!`. Only a single existing UI test was effected, with the result being an objective improvement.

~~I have left the method name and various other references to the Levenshtein algorithm untouched, as the exact manner in which the edit distance is calculated should not be relevant to the caller.~~

r? @estebank

@rustbot label +A-diagnostics +C-enhancement

jhpratt committed:
Use restricted Damerau-Levenshtein algorithm

jhpratt committed:
Add test for `println!` typo

est31 commented on compiler/rustc_span/src/lev_distance.rs:
```diff
@@ -1,49 +1,104 @@
-            } else {
-                dcol[j + 1] = cmp::min(current, next);
-                dcol[j + 1] = cmp::min(dcol[j + 1], dcol[j]) + 1;
+    // Variable length arrays are not yet implemented.
```
In C there is advice against using VLAs due to optimization and security reasons, with the kernel [dropping them completely](https://www.phoronix.com/news/Linux-Kills-The-VLA).

If you care about performance, you could use smallvecs here. I think if you put 16 characters as the limit that ought to be enough to not require allocations for 95% of literals out there.

In general this code should be rather cold though as it is only executed for diagnostics which the rustc compiler doesn't emit *that* many.

jhpratt replied:
I was not aware of that advice. I'll remove that comment.

This code should definitely be cold, which is why I didn't focus too much on eliminating the initial allocation of `Vec<char>`. But I was bored, so I figured let's do something useful.

`SmallVec` is something I considered. With the suggested changes elsewhere, I think it'll have minimal impact, though.

est31 commented:
Great PR, when reading rustc's suggested names, I always felt like there was room to improve.

cargo also uses the same distance algorithm, stored in `src/cargo/util/lev_distance.rs`. Maybe after this PR is merged one could think about filing a PR there as well.

steffahn commented:
The documentation of `lev_distance_with_substrings` still talks about ordinary “Levenshtein distance” even though it calls the `lev_distance` function whose implementation this PR changes.

tmiasko commented on compiler/rustc_span/src/lev_distance.rs:
```diff
@@ -1,49 +1,104 @@
+    loop {
+        match a.split_first().zip(b.split_first()) {
+            Some(((a_char, a_rest), (b_char, b_rest))) if a_char == b_char => {
+                a = a_rest.to_vec();
+                b = b_rest.to_vec();
```
Declare a and b as slices to avoid allocation here, e.g.,:

```rust
    let a = a.chars().collect::<Vec<_>>();
    let mut a = a.as_slice();
```

steffahn replied:
This could even use temporary lifetime extension
```rs
let mut a = &a.chars().collect::<Vec<_>>()[..];
```
though this might be considered less nice/idiomatic, depending on who you ask.

jhpratt replied:
I certainly felt like I was missing something here!

jhpratt commented:
@steffahn I'm thinking about just changing it to be a generic "edit distance" reference. There's no reason the caller should care about the exact algorithm, after all.

jhpratt committed:
Make public API, docs algorithm-agnostic

jhpratt committed:
Reduce limit on `macro_rules!` diagnostic

jhpratt commented:
Addressed all comments so far. I moved the functions and modules to correspond to a general "edit distance" rather than Levenshtein or Damerau-Levenshtein. In doing so I left a comment about the current implementation for anyone reading to know what it is.

One thing I ran across in doing this was a diagnostic for `macro_rules` typos. It had a limit of 3, which I've reduced to 2. I did this because the original issue was regarding a transposition, which changes the distance from 2 to 1 with this PR.

steffahn commented:
Just so I understand your comment correctly, you are saying that a limit of 3 instead of 2 for macros would produce only (or mostly) unhelpful additional suggestions, thus 2 is better?

jhpratt commented:
Note: this isn't for _macros_, it's a special diagnostic for `macro_rules` (as it's a contextual keyword). The original issue is #91227, with #91337 being the original implementation. The issue would be solved with 2 (current) or 1 (with this PR), but one was added to that.

I don't have strong feelings on it either way; it merely seemed appropriate.

steffahn commented:
Ah, thanks for the clarification and links. I only wanted to understand what this was about.

tmiasko commented:
Apart from the new UI test, can you also add unit tests that would distinguish the implementation from the Levenshtein distance, and unrestricted Damerau–Levenshtein distance?

jhpratt committed:
Add test for precise algorithm used

jhpratt commented:
@tmiasko Done.

tmiasko commented:
@bors r+

bors commented:
:pushpin: Commit ab4c0dd137f064b9a141d1341c546170b1c7272b has been approved by `tmiasko`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tmiasko ab4c0dd137f064b9a141d1341c546170b1c7272b -->
<!-- homu: {"type":"Approved","sha":"ab4c0dd137f064b9a141d1341c546170b1c7272b","approver":"tmiasko","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
